### ISSUE TITLE
PngEncoder Sws Scale

### ISSUE THREAD

**Issue by muratugureminoglu** [ant-media/Ant-Media-Server#2801]
### Environment

* Operating system and version: Ubuntu 18.04
* Java version:  Java11
* Ant Media Server version: Ant Media Server Enterprise 2.2.1

### Logs

```
Current thread (0x00007fc2bc014800):  JavaThread "vert.x-worker-thread-15" [_thread_in_native, id=2838, stack(0x00007fc2ab2f3000,0x00007fc2ab3f4000)]

Stack: [0x00007fc2ab2f3000,0x00007fc2ab3f4000],  sp=0x00007fc2ab3f1e28,  free space=1019k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libswscale.so.5+0x6d802]

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  org.bytedeco.ffmpeg.global.swscale.sws_scale(Lorg/bytedeco/ffmpeg/swscale/SwsContext;Lorg/bytedeco/javacpp/PointerPointer;Lorg/bytedeco/javacpp/IntPointer;IILorg/bytedeco/javacpp/PointerPointer;Lorg/bytedeco/javacpp/IntPointer;)I+0
j  io.antmedia.enterprise.preview.PngEncoder.writeFrameInternal(Lorg/bytedeco/ffmpeg/avutil/AVFrame;IJ)Z+202
j  io.antmedia.enterprise.adaptive.base.VideoEncoder.writeFrame(Lorg/bytedeco/ffmpeg/avutil/AVFrame;IJ)Z+26
j  io.antmedia.enterprise.adaptive.base.VideoEncoder.writeRawVideo([BI[IIIJIJ)Z+151
j  io.antmedia.enterprise.adaptive.WebRTCEncoderAdaptor.writeRawVideo([B[IIIIJJ)V+200
j  io.antmedia.enterprise.adaptive.WebRTCEncoderAdaptor.writeYuvPlanes([Ljava/nio/ByteBuffer;[IIIJJ)V+137
j  io.antmedia.enterprise.adaptive.WebRTCEncoderAdaptor.prepareAndEncodeBuffer(JJLorg/webrtc/VideoFrame$Buffer;)V+101
j  io.antmedia.enterprise.adaptive.WebRTCEncoderAdaptor.encodeVideoRunner()V+100
j  io.antmedia.enterprise.adaptive.WebRTCEncoderAdaptor.lambda$onAddStream$8(Lio/vertx/core/Promise;)V+1
j  io.antmedia.enterprise.adaptive.WebRTCEncoderAdaptor$$Lambda$385.handle(Ljava/lang/Object;)V+8
J 10640 c1 io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(Lio/vertx/core/spi/metrics/PoolMetrics;Ljava/lang/Object;Lio/vertx/core/Handler;Lio/vertx/core/Handler;)V (153 bytes) @ 0x00007fc4ce9bca94 [0x00007fc4ce9bc4e0+0x00000000000005b4]
J 10639 c1 io.vertx.core.impl.ContextImpl$$Lambda$353.run()V (24 bytes) @ 0x00007fc4ce9b65fc [0x00007fc4ce9b6540+0x00000000000000bc]
J 10710 c1 io.vertx.core.impl.TaskQueue.run()V (108 bytes) @ 0x00007fc4cea12e0c [0x00007fc4cea12900+0x000000000000050c]
J 10709 c1 io.vertx.core.impl.TaskQueue$$Lambda$187.run()V (8 bytes) @ 0x00007fc4ce77f9dc [0x00007fc4ce77f940+0x000000000000009c]
j  java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V+92 java.base@11.0.9.1
j  java.util.concurrent.ThreadPoolExecutor$Worker.run()V+5 java.base@11.0.9.1
j  io.netty.util.concurrent.FastThreadLocalRunnable.run()V+4
j  java.lang.Thread.run()V+11 java.base@11.0.9.1
v  ~StubRoutines::call_stub
```

[ant-media-server.log.zip](https://github.com/ant-media/Ant-Media-Server/files/5719197/ant-media-server.log.zip)

**Comment by muratugureminoglu** [comment]
[ant-media-server.log - 2020-12-28T134147.043.zip](https://github.com/ant-media/Ant-Media-Server/files/5746974/ant-media-server.log.-.2020-12-28T134147.043.zip)

**Comment by MarianoFacundoArch** [comment]
Same problem here causing random server restarts. Need this fixed asap!

**Comment by mekya** [comment]
Unfortunately, we cannot reproduce this problem. Do you have a reproduce scenario or server logs?

**Comment by MarianoFacundoArch** [comment]
Yes, of course. This is happening to us literally over 20 times a day per server. We have a cluster of 14 servers and is happening 20 times daily. Please tell me what information do you need so I can provide you everything.
We were almost to stop using the service in production, as this makes us drop all the incoming streams from the server it happens to.

The reproduce scenario is:
Start webrtc stream (we get this notified by the webhook)
As soon as we receive the webhook, we republish to a RTMP server running in Nimble

This happens a lot of times, as we do this for a lot of users. After some time, servers start restarting themselves with that exact same error ( io.antmedia.enterprise.preview.PngEncoder.writeFrameInternal ) .
Tell me how I can contact you directly so I can give you further information if needed. But we need this solved really ASAP, we have this in production and we get over 15 restarts per day per server dropping all the connections

**Comment by mekya** [comment]
Hi @MarianoFacundoArch ,

Quick workaround is disable png generation by putting 
```
settings.previewGenerate=false
```
to the `/usr/local/antmedia/webapps/WebRTCAppEE/WEB-INF/red5-web.properties` file
If you use `LiveApp` please replace `WebRTCAppEE` with `LiveApp` in the path above. 

For permanent solution, do you reproduce the problem? We don't reproduce the problem with our devices and our instances.

**Comment by MarianoFacundoArch** [comment]
You mean if I reproduce the problem after doing such modification?

**Comment by mekya** [comment]
Hi @MarianoFacundoArch ,
After this modification, your problem is likely fixed. 

I mean, do you reproduce the the problem before doing that modification? 
I mean just let me know the scenario and I'll reproduce the problem in order to fix

**Comment by MarianoFacundoArch** [comment]
Mekya, I am changing the settings on all my servers now to test if that hotfix fixes the issue, I will keep you informed about that.
Regarding reproducing it without any modification, yes, I can, as it happens to us around ~10 times PER SERVER PER DAY! I reproduce it all day long. We just receive Webrtc streams (around 20 per server) and republish them by RTMP republishing to a nimble server. Something important here: We tried both, with and without adaptive bitrate, and it happened all the same.

**Comment by MarianoFacundoArch** [comment]
To also comment, this happened when we did not have any cluster too.

This shows us that this might happen when:
Multiple WebRTC sterams + republishing (wait random time) --> Maybe when an user looses connectivity out of a sudden? 

I am testing the new hotfix by the way and will keep you informed on that too

**Comment by MarianoFacundoArch** [comment]
I can confirm that changing that setting has stopped the problem from happening, I will keep an eye on this. How can we track what's happening to properly fix it without disabling the setting?

**Comment by mekya** [comment]
Hi @MarianoFacundoArch ,

It's great to hear that it's working for you even if it's a workaround solution. 

if the problem is this one
```
j  org.bytedeco.ffmpeg.global.swscale.sws_scale(Lorg/bytedeco/ffmpeg/swscale/SwsContext;Lorg/bytedeco/javacpp/PointerPointer;Lorg/bytedeco/javacpp/IntPointer;IILorg/bytedeco/javacpp/PointerPointer;Lorg/bytedeco/javacpp/IntPointer;)I+0
```
It's not related to cluster, it's not related to RTMP publishing either. It's about a memory issue in preview generation. 

- How do you publish the stream? 
- Publishing stream is under your control or you have users publishing from their web browsers? 


It will be great if I can have the exact reproduce scenario. It'll help a lot to understand the problem.

**Comment by MarianoFacundoArch** [comment]
The use case is this:
We receive WebRTC streams pushed from our users in their web browsers. We have an API webhook that gets notified everytime a stream starts, so as soon as ant calls our hook, we call ant back telling to republish to a specific RTMP server that runs on nimble.

**Comment by mekya** [comment]
Thank you for the case. 

Can we learn the device and the browser of the user that cause this problem?

I can find out the related stream ids. I mean if you have a kind of record which stream id belongs to which user, we may find the user and ask him to his device and browser.

**Comment by MarianoFacundoArch** [comment]
I don't have that data.
Sorry.
I have no way of which stream caused the problem, as the server just restarts.
That setting you told me to modify, is still included in 2.3.0? 
As if we can't modify and disable that thing that is bugged we can't use the software propperly.
BTW, How can we fix this forever?

**Comment by mekya** [comment]
H @MarianoFacundoArch ,

>I have no way of which stream caused the problem, as the server just restarts.
>BTW, How can we fix this forever?

I can find the stream id that causes the crash. Can you find which user has that stream id?Then we may learn their device name and browser?

>That setting you told me to modify, is still included in 2.3.0?

Yes sure. This setting will be always available because turning on/off preview generation is something configurable.

**Comment by duyduy** [comment]
Here are the steps I reproduce the issue: https://www.youtube.com/watch?v=r2B4d-4XteE&ab_channel=duyduy. 
I'm using WebRTCiOSReferenceProject_15_Jan_202 and Ant server : Enterprise Edition 2.2.1 20201029_2042. 

Hope that help you can reproduce the issue .

**Comment by mekya** [comment]
Thank you for your report, I'll absolutely check it.

**Comment by mekya** [comment]
Thank you. Your device is up to date, isn't it ? @duyduy

**Comment by duyduy** [comment]
It's iOS 14.4 .

**Comment by mekya** [comment]
Thank you guys (@MarianoFacundoArch  @duyduy ) for your contributions. You help me a lot to spot the issue.  

Last night we've fixed this issue and it is going to be available 2.3.0 that we plan to release today.

Thank you so much
